src/Persistence_matrix/include/gudhi/matrix.h

```diff
@@ -0,0 +1,2106 @@
+   * @ref Matrix(const std::function<bool(pos_index,pos_index)>&, const std::function<bool(pos_index,pos_index)>&)
+   * for more information about the comparators.
+   *
+   * @tparam Boundary_type Range type for @ref cell_rep_type ranges. Assumed to have a begin(), end() and size() method.
+   * @param orderedBoundaries Vector of ordered boundaries in filtration order. Indexed continously starting at 0.
+   * @param birthComparator Method taking two @ref PosIdx indices as parameter and returns true if and only if the first
+   * face is associated to a bar with strictly smaller birth than the bar associated to the second one.
+   * @param deathComparator Method taking two @ref PosIdx indices as parameter and returns true if and only if the first
+   * face is associated to a bar with strictly smaller death than the bar associated to the second one.
+   * @param characteristic Characteristic of the coefficient field. Has to be specified if
+   * @ref PersistenceMatrixOptions::is_z2 is false. Default value is 11.
+   * Ignored if @ref PersistenceMatrixOptions::is_z2 is true.
+   */
+  template <class Boundary_type = boundary_type>
+  Matrix(const std::vector<Boundary_type>& orderedBoundaries, 
+         const std::function<bool(pos_index,pos_index)>& birthComparator,
+         const std::function<bool(pos_index,pos_index)>& deathComparator, 
+         characteristic_type characteristic = 11);
+  /**
+   * @brief Constructs a new empty matrix and reserves space for the given number of columns.
+   * Only available when those comparators are necessary.
+   *
+   * That is, when **all** following options have following values:
+   *   - @ref PersistenceMatrixOptions::is_of_boundary_type = false
+   *   - @ref PersistenceMatrixOptions::has_vine_update = true
+   *   - @ref PersistenceMatrixOptions::has_column_pairings = false
+   *
+   * See description of
+   * @ref Matrix(const std::function<bool(pos_index,pos_index)>&, const std::function<bool(pos_index,pos_index)>&)
+   * for more information about the comparators.
+   *
+   * @param numberOfColumns Number of columns to reserve space for.
+   * @param birthComparator Method taking two @ref PosIdx indices as parameter and returns true if and only if the first
+   * face is associated to a bar with strictly smaller birth than the bar associated to the second one.
+   * @param deathComparator Method taking two @ref PosIdx indices as parameter and returns true if and only if the first
+   * face is associated to a bar with strictly smaller death than the bar associated to the second one.
+   * @param characteristic Characteristic of the coefficient field. If not specified and
+   * @ref PersistenceMatrixOptions::is_z2 is false, the characteristic has to be set later with the use of
+   * @ref set_characteristic before calling for the first time a method needing it.
+   * Ignored if @ref PersistenceMatrixOptions::is_z2 is true.
+   */
+  Matrix(unsigned int numberOfColumns, 
```

Comment:
I wanted to use a native type here to indicate that `numberOfColumns` is not something stored (and therefore doesn't really need to be optimized in size). And the chances are not very high that someone wants to use more than max unsigned int columns. Otherwise, `index` would be the right type to use.
I am more bothered by the fact that I use `unsigned int` in one constructor and `int` in another. This makes no sense.

Reply:
> I wanted to use a native type here to indicate that `numberOfColumns` is not something stored (and therefore doesn't really need to be optimized in size).

As a user, that's not the message I understand when I read `unsigned int`, almost the opposite. `std::size_t` or `std::intmax_t` (or maybe `ptrdiff_t`) would be closer.

> the chances are not very high that someone wants to use more than max unsigned int columns.

"640K [of RAM] ought to be enough for anybody" :wink:
A few years after starting Gudhi, we received a bug report because the code failed on a simplicial complex with 3 billion simplices. Sure, that required an awful lot of memory to reproduce, but I don't see a good reason to hardcode a limit of 4 billions (the default options can still limit the computation to 4G, that's fine).